### DISCUSSION TITLE
magick++ is performing HTTP request when parsing SVG.

### DISCUSSION THREAD

**Discussion by StephenLynx** [ImageMagick/ImageMagick#4817] [General]
I'm using the function readImages, as seen here: https://gitgud.io/LynxChan/LynxChan/-/blob/master/src/be/native/imageThumb.cpp#L19

The file is the following: https://pastebin.com/6sZyY6rM

And I got a report on my project that this causes a HTTP request to be performed on the URL set in xlink. Is this intentional? Can it be disabled?

**Comment by urban-warrior** [comment]
See https://imagemagick.org/script/security-policy.php.

**Comment by StephenLynx** [comment]
Thanks, we were able to prevent this specific issue I posted here, but now there's something about XMP metadata on binary image formats.

`XXE through injection of a Entity payload in the XMP metadata of a TIFF file. The server parsed the code <?xpacket begin='ï»¿' id='W5M0MpCehiHzreSzNTczkc9d'?><!DOCTYPE root [ <!ENTITY xxe SYSTEM "http://example.org/x"> ]> [...] <pdf:Keywords>&xxe;</pdf:Keywords> which resulted in a SSRF.
`
`XXE through injection of a Entity payload in the XMP metadata of a JPEG file. The server parsed the code <?xpacket begin='ï»¿' id='W5M0MpCehiHzreSzNTczkc9d'?><!DOCTYPE root [ <!ENTITY xxe SYSTEM "http://example.org/x"> ]> [...] <pdf:Keywords>&xxe;</pdf:Keywords> which resulted in a SSRF. `

**Comment by urban-warrior** [comment]
You can strip XMP profiles from an image after it is read.  We could also add an option to skip over the XMP profile of an image as it is read.  We would need details of this potential vulnerability plus one or two example image exploits before we can offer any additional advice.

**Reply by StephenLynx** [reply]
Stripping XMP profiles AFTER reading is pointless because the file was already read and the exploit already executed. The guy gave me this file: https://gitgud.io/LynxChan/LynxChan/uploads/a6b5b1fe1f4081131a883d90ed08b6cb/file.jpg

**Reply by urban-warrior** [reply]
We need to reproduce an exploit before we can mitigate the potential vulnerability.  `file.jpg` does not include an XMP profile.  It does include an EXIF profile.  We ran our memory checkers on the file and it did not return any exploits-- no remote access was initiated (per `strace`), no memory was stepped on (per `valgrind`), and `magick` returned with a proper exception: 

```
magick: Corrupt JPEG data: 67 extraneous bytes before marker 0xdb `file.jpg' @ warning/jpeg.c/JPEGWarningHandler/402.
magick: Quantization table 0x00 was not defined `file.jpg' @ error/jpeg.c/JPEGErrorHandler/347.
```

Provide a method to reproduce the exploit and we will provide a work-around or patch to mitigate the issue.

**Reply by StephenLynx** [reply]
We didn't use "magick", we are using the c++ libraries. Try running this program:
`#define MAGICKCORE_HDRI_ENABLE 0
#define MAGICKCORE_QUANTUM_DEPTH 16

#include <Magick++.h>

int main(int argc, char **argv)
{
  std::list < Magick::Image > frameList;
  
  if(argc < 2){
    puts("Inform the file path as the first argument");
  } else {
    readImages(&frameList, argv[1]);
  }
  
}
`

And see if it makes a HTTP request.

**Reply by ipsec403** [reply]
I can't reproduce it with the example above either. The issue is more complicated than I've previously thought. I need to do more testing before I can say for certain what caused the XXE payload to trigger in `readImages`.

**Reply by ipsec403** [reply]
@urban-warrior [PoC.zip](https://github.com/ImageMagick/ImageMagick/files/8025934/PoC.zip)

**Step 1**: `mkdir test && cd test && python3 -m http.server`

```
Serving HTTP on 0.0.0.0 port 8000 (http://0.0.0.0:8000/) ...
```

**Step 2**: `make && ./imTest file1 file2` 

```
file1
file2
error : Operation in progress
I/O warning : failed to load HTTP resource
```

**Step 3**: Now we can see one GET request for `file2`

```
127.0.0.1 - - [08/Feb/2022 XX:XX:XX] code 404, message File not found
127.0.0.1 - - [08/Feb/2022 XX:XX:XX] "GET /x HTTP/1.0" 404 -
```

Is that expected because the exception `Magick::ErrorCoder` was thrown? What can we do about it?

**Additional information**:

libmagick++-dev: `6.9.10.23+dfsg-2.1ubuntu11.4`

`/etc/ImageMagick-6/policy.xml`:
```
<policymap>
  <policy domain="resource" name="memory" value="256MiB"/>
  <policy domain="resource" name="map" value="512MiB"/>
  <policy domain="resource" name="width" value="16KP"/>
  <policy domain="resource" name="height" value="16KP"/>
  <policy domain="resource" name="area" value="128MB"/>
  <policy domain="resource" name="disk" value="1GiB"/>
  <policy domain="path" rights="none" pattern="@*"/>
  <policy domain="coder" rights="none" pattern="PS" />
  <policy domain="coder" rights="none" pattern="PS2" />
  <policy domain="coder" rights="none" pattern="PS3" />
  <policy domain="coder" rights="none" pattern="EPS" />
  <policy domain="coder" rights="none" pattern="PDF" />
  <policy domain="coder" rights="none" pattern="XPS" />
  <policy domain="coder" rights="none" pattern="EPHEMERAL" />
  <policy domain="coder" rights="none" pattern="URL" />
  <policy domain="coder" rights="none" pattern="HTTP" />
  <policy domain="coder" rights="none" pattern="HTTPS" />
  <policy domain="coder" rights="none" pattern="MVG" />
  <policy domain="coder" rights="none" pattern="MSL" />
  <policy domain="coder" rights="none" pattern="TEXT" />
  <policy domain="coder" rights="none" pattern="SHOW" />
  <policy domain="coder" rights="none" pattern="WIN" />
  <policy domain="coder" rights="none" pattern="PLT" />
  <policy domain="coder" rights="none" pattern="FTP" />
</policymap>
```

**Reply by urban-warrior** [reply]
We're investigating why the exception was not caught, we'll get back to you.  It is certainly thrown per the command-line:

```
/usr/bin/convert file2 null:
convert: CorruptImageProfile `file2' (XMP) @ warning/profile.c/ValidateXMPProfile/1701.
convert: Corrupt JPEG data: 28 extraneous bytes before marker 0xdb `file2' @ warning/jpeg.c/JPEGWarningHandler/389.
convert: Quantization table 0x00 was not defined `file2' @ error/jpeg.c/JPEGErrorHandler/338.
```

**Reply by ipsec403** [reply]
FWIW, we just confirmed that the PoC doesn't work on CentOS 7 and Rocky 8. So far at least Debian 10 and Ubuntu 20.04 are confirmed to be vulnerable.

**Reply by ipsec403** [reply]
Another question: How to specify the security policy with a parameter in the C++ code or via the build options?

**Reply by urban-warrior** [reply]
With IM7 you can use https://imagemagick.org/api/Magick++/classMagick_1_1SecurityPolicy.html,  For IM6, you may need to set environment variables @ https://imagemagick.org/script/resources.php#environment or add a custom security policy in your local path, .e.g., `~/.config/ImageMagick/policy.xml` and verify with this command: `convert -list policy`.

**Reply by ipsec403** [reply]
Thanks. Now our plan is to use `MAGICK_CONFIGURE_PATH` and only provide the `policy.xml` file within that path.

**Comment by StephenLynx** [comment]
Update: centos 7 and rocky 8 are not vulnerable, debian 10 and some ubuntu version are.

**Comment by ipsec403** [comment]
@urban-warrior A LynxChan user wrote this in our internal forums:

--------------------------------------

Debian 10 and 11 are both vulnerable when used with `libmagic++-dev`. When ImageMagick6 is compiled by sources, it depends on the installed build dependencies. In a test with just libjpeg-dev, libpng-dev, libwebp-dev, libgif-dev, libopenjp2-7-dev, libheif-dev, libtiff-dev, libbrotli-dev, libbz2-dev and libexif-dev installed while configuring and installing ImageMagick, PoC.zip could not be reproduced. In a try with all of Debian's standard build dependencies (`apt-get -s build-dep imagemagick`) installed, PoC.zip could be reproduced (doesn't matter if with policy.xml rules or not). 

<details><summary>Vulnerable version - identify --version - Click to expand</summary>
Version: ImageMagick 6.9.12-39 Q16 x86_64 2022-02-05 https://imagemagick.org
...
Delegates (built-in): bzlib djvu fontconfig freetype heic jbig jng jp2 jpeg lcms lqr ltdl lzma openexr pangocairo png tiff webp wmf x xml zlib
</details>

<details><summary>Invulnerable version - identify --version - Click to expand</summary>
Version: ImageMagick 6.9.12-39 Q16 x86_64 2022-02-05 https://imagemagick.org
...
Delegates (built-in): bzlib heic jbig jng jp2 jpeg ltdl lzma png tiff webp zlib 
</details>

So it depends probably on some of those shown delegates.

--------------------------------------

Maybe that helps.

**Reply by urban-warrior** [reply]
ImageMagick does not do anything with the XMP profile.  It treats it as a blob.  It does pass the blob to `libxml2` for validation.  Perhaps certain releases of the `libxm2` library is the source of the vulnerability.  However, add a define, `-define profile:skip='*'` to skip consuming the profiles.  That likely will resolve the issue, assuming your release of ImageMagick supports the profile define.

**Reply by ipsec403** [reply]
> add a define, `-define profile:skip='*'` to skip consuming the profiles

But isn't that a command-line option for the `magick` binary? We want to solve it for our specific usage of `libmagic++-dev` as demonstrated in `PoC.zip`.

**Reply by urban-warrior** [reply]
For Magick++, use `defineSet`.

**Reply by ipsec403** [reply]
Thanks, but as discussed above, that's already too late because the payload has been read at this point. As I understand it, @StephenLynx doesn't want to use the `magick` binary for performance reasons. Adding an option to skip over the XMP profile of an image as it is read would probably be the best course of action.
```
 std::list < Magick::Image > frameList;
 readImages(&frameList, argv[1]);
```

**Reply by ipsec403** [reply]
We've found a solution in regards to the security policy. Unsurprisingly in retrospect, the pattern `SVG` should have been disallowed in addition... Although if somebody wanted to read an SVG file via `readImages`, the vulnerability would persist.

**Reply by urban-warrior** [reply]
SVG files are not read (or written) if the security policy prohibits it.   For JPEG and TIFF, use the `profile:skip` define to prevent the profiles from being consumed.  Given these two guards, if the vulnerability persists, it should not be within ImageMagick, but instead may be exposed elsewhere in your workflow.
